tornado/gen.py

```diff
@@ -1236,17 +1242,14 @@ def cb():
             self.io_loop.add_callback(self.run)
             return False
         elif not self.future.done():
-            def inner(f):
-                # Break a reference cycle to speed GC.
-                f = None  # noqa
-                self.run()
-            self.io_loop.add_future(
-                self.future, inner)
+            def inner(f, run = wrap(self.run), io_loop = self.io_loop):
+                io_loop.add_callback(run)
+            future_add_done_callback(self.future, inner)
```

Comment:
This doesn't appear to do the same. Before, 'inner' called run directly, whereas now it queues a callback that will then call run. Is that intentional ?

Reply:
Well, if it did the same, it would not achieve any improvement.

This is not a micro-optimization, it's a different way of achieving the same that doesn't hold a reference to `f` for longer than absolutely necessary. The original code would hold a reference to it while `run()` was running. The reference is hidden in the caller of the callback.

If you check what `add_future` does, this code does the same but using an interface that doesn't pass the future to the callback (ie: `add_callback` instead of `add_future`), so there's no lingering reference.